File: AndroidUsingJava/DesignPatterns/newMVP2/app/src/main/java/com/example/newmvp/db/ProductDataBase.java
package com.example.newmvp.db;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import com.example.newmvp.model.Product;

@Database(entities = {Product.class},version = 1 )
public abstract class ProductDataBase extends RoomDatabase {
    private static ProductDataBase DBinstance = null;
    public abstract ProductDAO getProductDAO();
    public static synchronized ProductDataBase getInstance(Context context){
        if(DBinstance == null){
            DBinstance = Room.databaseBuilder(context.getApplicationContext(),ProductDataBase.class,"product").build();
        }
        return DBinstance;
    }

}

File: AndroidUsingKotlin/Session_7/MVVM/app/src/main/java/com/example/mvvm/DB/ProductDataBase.kt
package com.example.mvvm.DB

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import com.example.mvvm.Product

@Database(entities = arrayOf(Product::class), version = 1)
abstract class ProductDatabase : RoomDatabase() {
    abstract fun getproductDao(): ProductDao
    companion object {
        @Volatile
        private var INSTANCE: ProductDatabase? = null

        fun getInstance(context: Context): ProductDatabase {
            return INSTANCE ?: synchronized(this) {
                val instance = Room.databaseBuilder(
                    context.applicationContext,
                    ProductDatabase::class.java,
                    "product_database"
                ).build()
                INSTANCE = instance
                instance
            }
        }
    }
}